### ISSUE TITLE
Azkaban job configuration per environment

### ISSUE THREAD

**Issue by vladislav-sidorovich** [azkaban/azkaban#1545]
Hi guys,

is it possible to configure Azkaban to use different property files for job configuration on each environment like dev, stage, prod and so on? Like described here https://stackoverflow.com/questions/46939571/azkaban-job-configuration-per-environment

**Comment by reallocf** [comment]
Hey @vladislav-sidorovich,

Looking at your stackoverflow example, I'm not sure how well the nested ${${}} will be handled. Could you set the db configuration in the system.properties file itself?

For example:
```
system.properties
myFlow/
   foo.job

#system.properties in dev environment
db=localhost:2181

#system.properties in prod environment
db=aws:port

#foo.job
some command --db ${db}
```
That's how we handle stuff here - alternating the configs themselves for the dev/staging/prod environments so individual flow/job doesn't have to worry about it.

Does that make sense?

Thanks!
Charlie

**Comment by vladislav-sidorovich** [comment]
Hi @reallocf  thank you for answer!

The example on stackoverflow works :). And your example works as well.  And your example works as well. 

But the  idea is to provide few property files (local, test, stage, prod) from developer's side because all these files will keep in Git. And it's more useful to have few files with different values of properties than keep one file and change the value of properties each time.

**Comment by reallocf** [comment]
Well if your solution is working for you then feel free to go with that!

Another option would be to label each environment in a `common.properties` file in the `plugins/jobtypes/` directory in your executor or solo-server. Properties defined there will be picked up by all jobtypes on that Azkaban instance (like global variables affecting the scope of every job) and you could differ the label for each environment.

After doing that you could define in code using a hadoopJava job or something else that executes arbitrary code.

Small example:
```
# In built instance
azkaban-exec-server/ (or azkaban-solo-server)
   bin/
   conf/
   lib/
   executions/
   plugins/
      jobtypes/
         commonprivate.properties
         common.properties:
            env.LOCATION=prod

# In your zip file
system.properties
myFlow/
   flow.job:
      type=command (hadoopJava jobtype is preferable, but command works for a small demo)
      command=python test.py
   test.py:
      from os import environ
      if environ["LOCATION"] == prod:
         print("This is prod!")
```
In this way, users only have to think about defining things in code which may or may not be easier depending on the complexity of your files and who will end up using your Azkaban installation.

There are probably another 4 or 5 ways to get your desired outcome. I'd use whatever works best for you and your team :smile:

**Comment by vladislav-sidorovich** [comment]
Thank you very much.

**Comment by juhoautio** [comment]
@reallocf please consider this: it's common that users want to set environment-specific configurations in their own projects (zips) – so being able to set environment variables like env.LOCATION doesn't help there. How would you give an env-specific arg to your test.py? I don't see a way. The global config can't be changed by the users, and it's painful to add new properties in there.

----

We have implemented support for different environments in _our custom job types_.

It works like this.

Environment is set globally on the azkaban installation;
```
azkaban-exec-server/
   plugins/
      jobtypes/
         commonprivate.properties
         common.properties:
            environment=dev
```

When using our custom job types, the zip can be for example:

```properties
myFlow/
  myFlow.properties
  foo.job

# myFlow.properties
database=localhost:2181
dev.database=dev-db:2181
stage.database=some-host:2181
prod.database=another-host:2181

# foo.job
type=custom_type
some_arg=test value
prod.some_arg=real value
command=some_command --db ${database} "${some_arg}"
```

After resolving by env prefixes the result becomes =>

```properties
command=some_command --db dev-db:2181 "test value"
```

When our custom job types resolve props, they look for `${environment}.*` ie. `dev.*` in this case, and insert its value as the flat property, in this case `dev.database` -> `database`. This overrides a possibly existing non-prefixed value.

----

I'd be really interested to hear if anyone has better ideas.

I wonder if it would make sense for azkaban itself to support this pattern. But I also feel that it's a bit "magical" and maybe not so beautiful in that sense, but it seems to work well for us.

I know that in many cases the environment specific configurations have been chosen to live in separate files. But it's always a bore having to edit 3 different files when you want to tweak some special property.

To me "our way" seems more like "package by feature" ([which is good](http://www.javapractices.com/topic/TopicAction.do?Id=205)) while having env-specific files reminds me of "package by layer" (a mess).

IMHO, having separate files like `myFlow-dev.properties, myFlow-stage.properties, myFlow-prod.properties` is not so natural many times, especially when you just have a simple job that needs some value to be different in each env. See that `prod.some_arg` "override" in `foo.job` – if the property is specific to this job (and you have many such jobs that all need different values per environment), how would you accomplish that? You could maybe put the job file into a separate sub-folder and create something like `foo/foo-prod.properties` just to set that value.. It doesn't seem too appealing to me.

----

If we would be using the "out-of-the-box" solution `${${environment}.database}` it would more or less mean that we would need to prefix every property name with `${environment}.`. Too much hassle. And if you want to set a property that is implemented by a job type directly, for example Azkaban's own java job types' `classpath`, you have to always go via a reference. You can't just do `dev.classpath=something` but have to define separately `dev.classpath=something` + `classpath=${${environment}.classpath}` – so much more verbose and hard to follow.

----

By the way, one thing that can't be done in any way at the moment is setting env-specific number of `retries=` for a job. Or `dependencies=`.
